File: planctl/discovery.py
"""planctl multi-project discovery (fn-542).

Pure filesystem scan: given the configured ``roots`` (parent directories), walk
each root's **immediate children** and return those that contain a ``.planctl/``
directory. Mirrors jobctl's ``_discover_planctl_projects`` semantics —
immediate children only, **skip nested** ``.planctl/`` (agent worktrees live at
``<project>/.claude/worktrees/<id>/.planctl/`` and must not double-count).

No daemon, no devctl, no jobctl cache — planctl owns its own roots. Fail-soft: a
root that doesn't exist (or can't be listed) is skipped, not an error.
"""

from __future__ import annotations

from dataclasses import dataclass, field
from pathlib import Path

from planctl.config import load_roots


@dataclass(frozen=True)
class ResolveResult:
    """Outcome of ``resolve_epic_globally`` (fn-600).

    Cwd-then-global epic-id resolution distinguishes the three observable
    outcomes callers need to branch on without re-parsing or re-scanning:

    - **Resolved**: ``project_path`` is the owning project root and
      ``epic_path`` is the ``.planctl/epics/<id>.json`` path inside it.
      ``resolved_id`` is the full slug id of the matched epic (equal to the
      input for a slug input, the canonical slug for a number-only ``fn-N``
      input). ``owners`` is empty.
    - **Not found**: id matches no project's on-disk epic. ``project_path``,
      ``epic_path``, and ``resolved_id`` are all ``None``. ``owners`` is empty.
    - **Ambiguous**: id appears in two or more projects (legacy dup state —
      ``_check_global_name_unique`` post-fix prevents new ones, but historical
      collisions may exist). ``project_path``, ``epic_path``, and
      ``resolved_id`` are ``None``; ``owners`` lists every project carrying the
      id. Callers surface a ``dep_ambiguous_id`` error envelope.

    Hard contract: ambiguous-id is an error, not a silent pick. Do NOT follow
    ``scan_epic_ids_global``'s "last-walked wins" comment as resolver
    semantics; that's for human-readable error messages in dup-detection
    paths only.

    Number-only inputs (fn-600 follow-up): a bare ``fn-N`` (no slug) resolves
    by exact epic-integer equality via :func:`planctl.ids.parse_id` — never
    string-prefix matching, so ``fn-1`` never matches ``fn-10``. The resolver
    returns the matched epic's FULL slug id in ``resolved_id`` so the
    dep-writing path can normalize the persisted edge to canonical form.
    """

    project_path: Path | None
    epic_path: Path | None
    owners: list[Path] = field(default_factory=list)
    resolved_id: str | None = None

    @property
    def resolved(self) -> bool:
        """True iff the id resolved to exactly one project."""
        return self.project_path is not None

    @property
    def ambiguous(self) -> bool:
        """True iff two or more projects carry the id."""
        return len(self.owners) > 1


def resolve_epic_globally(
    epic_id: str, roots: list[Path] | None = None
) -> ResolveResult:
    """Resolve an epic id cwd-then-global, distinguishing not-found from ambiguous.

    Resolution order (shared cwd-then-roots lookup):

    1. **Cwd**: if the current working directory is a planctl project AND its
       ``.planctl/epics/<epic_id>.json`` exists, return it. This preserves the
       single-repo workflow (no configured ``roots`` required) and is also the
       canonical short-circuit when the id lives in the parent epic's own
       project — by far the common case.
    2. **Roots discovery**: scan configured (or given) ``roots`` for projects
       whose ``.planctl/epics/<epic_id>.json`` exists. Exactly one match →
       resolved. Zero → not found. Many → ambiguous (legacy dup state).

    The cwd-first short-circuit must NOT count toward the ambiguity check —
    if the id resolves locally we never look further. Callers that want the
    full multi-project view (e.g. integrity restamp building the global
    ``all_epic_ids`` set) use :func:`scan_epic_ids_global` directly.

    Fail-soft on roots discovery: if ``discover_projects()`` raises or yields
    nothing, the global step contributes no candidates — callers see a
    not-found result, never an exception. The cwd short-circuit still works
    in that degraded case (no configured ``roots``).
    """
    from planctl.project import find_project_root

    # 1. Cwd short-circuit: keeps single-repo workflows working without
    # configured roots, and is the common hot path when the dep lives in the
    # parent epic's own project.  ``find_project_root`` already swallows its
    # known failure modes (git not on PATH, not a git repo) and falls back to
    # ``Path.cwd()``; any further OSError (e.g. unreadable cwd) is the only
    # remaining failure surface here.
    cwd_root: Path | None
    try:
        cwd_root = find_project_root()
    except OSError:
        cwd_root = None
    if cwd_root is not None and (cwd_root / ".planctl").is_dir():
        cwd_hit = _match_epic_in_project(cwd_root, epic_id)
        if cwd_hit is not None:
            resolved_id, cwd_epic = cwd_hit
            return ResolveResult(
                project_path=cwd_root,
                epic_path=cwd_epic,
                resolved_id=resolved_id,
            )

    # 2. Roots discovery. Fail-soft: an unconfigured / unwritable / empty
    # discovery yields zero candidates, equivalent to "not found".  Each match
    # carries the project root and the matched epic's FULL slug id so a
    # number-only ``fn-N`` input normalizes to canonical form.
    try:
        matches = _find_epic_matches(epic_id, roots)
    except Exception:
        matches = []

    # Filter the cwd path out of the candidates so we don't double-count it as
    # ambiguous when both cwd and discovery surface the same project (e.g.
    # cwd is itself under a configured root but its on-disk epic JSON was
    # absent above, which would only happen on a freshly-deleted file race —
    # belt-and-suspenders).
    if cwd_root is not None:
        try:
            cwd_real = cwd_root.resolve()
        except Exception:
            cwd_real = cwd_root
        matches = [m for m in matches if m[0].resolve() != cwd_real]

    if not matches:
        return ResolveResult(project_path=None, epic_path=None)
    if len(matches) == 1:
        owner, resolved_id = matches[0]
        return ResolveResult(
            project_path=owner,
            epic_path=owner / ".planctl" / "epics" / f"{resolved_id}.json",
            resolved_id=resolved_id,
        )

    # Ambiguous: legacy dup state, or a number-only ``fn-N`` whose integer
    # matches an epic in two or more projects. Callers surface
    # dep_ambiguous_id; the resolver does NOT silently pick a winner.
    return ResolveResult(
        project_path=None, epic_path=None, owners=[m[0] for m in matches]
    )


def _match_epic_in_project(project: Path, epic_id: str) -> tuple[str, Path] | None:
    """Resolve *epic_id* against one project, returning ``(full_id, epic_path)``.

    A full slug input matches its own ``.planctl/epics/<epic_id>.json`` exactly.
    A number-only ``fn-N`` input matches the epic whose parsed ``epic_num``
    equals N by integer equality (never string-prefix: ``fn-1`` does not match
    ``fn-10``). Within one project the epic integer is unique by construction,
    so at most one file matches. Returns ``None`` when nothing matches.
    """
    from planctl.ids import parse_id

    epics_dir = project / ".planctl" / "epics"

    # Full-slug (or any id that exists as an exact filename): canonical exact
    # match, no scan needed. This is the hot path for slug inputs.
    exact = epics_dir / f"{epic_id}.json"
    if exact.exists():
        return epic_id, exact

    # Number-only ``fn-N``: match by epic-integer equality. A slug input that
    # didn't exist above falls through here and finds nothing (its task_num is
    # None and no bare-number file will equal a slug), so this stays safe.
    want_epic, want_task = parse_id(epic_id)
    if want_epic is None or want_task is not None:
        return None
    # Only treat the input as number-only when it has no slug tail — a slug
    # input that simply doesn't exist must NOT match a same-number epic.
    if epic_id != f"fn-{want_epic}":
        return None
    try:
        candidates = sorted(epics_dir.glob("fn-*.json"))
    except OSError:
        return None
    for candidate in candidates:
        stem = candidate.stem
        cand_epic, cand_task = parse_id(stem)
        if cand_task is not None:
            continue
        if cand_epic == want_epic:
            return stem, candidate
    return None


def _find_epic_matches(
    epic_id: str, roots: list[Path] | None = None
) -> list[tuple[Path, str]]:
    """Return ``(project_root, full_epic_id)`` for every project matching *epic_id*.

    Sibling of :func:`find_projects_with_epic` that also carries the matched
    epic's FULL slug id, so a number-only ``fn-N`` input normalizes to canonical
    form on the write side. Matching delegates to :func:`_match_epic_in_project`
    (exact-filename for a slug, integer-equality for a bare number).
    """
    matches: list[tuple[Path, str]] = []
    for project in discover_projects(roots):
        hit = _match_epic_in_project(project, epic_id)
        if hit is not None:
            matches.append((project, hit[0]))
    return matches


def find_projects_with_task(
    task_id: str, roots: list[Path] | None = None
) -> list[Path]:
    """Return discovered project roots whose ``.planctl/tasks/<task_id>.json`` exists.

    Scans the configured (or given) ``roots`` via :func:`discover_projects`, then
    filters to those holding the named task definition on disk. Used by ``claim``
    to resolve a task's owning project cwd-agnostically (fn-542 task .3).

    Returns absolute, deduplicated, sorted project roots (inherits the ordering
    guarantees of :func:`discover_projects`). Empty list when no project holds
    the task — the caller maps that to ``TASK_NOT_FOUND``.
    """
    matches: list[Path] = []
    for project in discover_projects(roots):
        if (project / ".planctl" / "tasks" / f"{task_id}.json").exists():
            matches.append(project)
    return matches


def find_projects_with_epic(
    epic_id: str, roots: list[Path] | None = None
) -> list[Path]:
    """Return discovered project roots whose ``.planctl/epics/<epic_id>.json`` exists.

    Sibling of :func:`find_projects_with_task` for epic-keyed lookups. Used by
    epic-keyed verbs (e.g. ``epic rm``) to resolve an epic's owning project
    cwd-agnostically when the operator invokes them from outside the repo
    whose ``.planctl/`` holds the epic. Accepts a number-only ``fn-N`` id
    (integer-equality match) as well as a full slug.
    """
    return [project for project, _full_id in _find_epic_matches(epic_id, roots)]


def discover_projects(roots: list[Path] | None = None) -> list[Path]:
    """Return planctl project directories under the given (or configured) roots.

    A project is an **immediate child** of a root that contains a ``.planctl/``
    directory. Nested ``.planctl/`` dirs (worktrees, vendored copies, archived
    snapshots) are intentionally NOT surfaced — only one level of children is
    scanned per root.

    Roots are deduplicated and missing / unlistable roots are skipped silently.
    Returned paths are absolute and deduplicated, sorted for determinism.
    """
    if roots is None:
        roots = load_roots()

    projects: list[Path] = []
    seen: set[str] = set()
    seen_roots: set[str] = set()

    for root in roots:
        root_key = str(root)
        if root_key in seen_roots:
            continue
        seen_roots.add(root_key)

        try:
            children = sorted(root.iterdir())
        except OSError:
            # Root doesn't exist or can't be listed — skip, not an error.
            continue

        for child in children:
            try:
                if not child.is_dir():
                    continue
                if not (child / ".planctl").is_dir():
                    continue
            except OSError:
                continue
            key = str(child)
            if key in seen:
                continue
            seen.add(key)
            projects.append(child)

    return sorted(projects)

File: plugins/plan/src/discovery.ts
// Multi-project discovery — the subset of planctl/discovery.py that claim needs.
//
// Pure filesystem scan: given the configured roots (parent directories), walk
// each root's IMMEDIATE children and return those holding a `.keeper/` data dir.
// Immediate children only — nested data
// dirs (agent worktrees) must not double-count. Fail-soft: a root that doesn't
// exist (or can't be listed) is skipped, not an error.

import { existsSync, readdirSync, realpathSync, statSync } from "node:fs";
import { join } from "node:path";

import { loadRoots } from "./config.ts";
import { parseId } from "./ids.ts";
import { classifyCwdVantage } from "./project.ts";
import {
  hasDataDir,
  resolveDataDir,
  resolveDataDirOrDefault,
} from "./state_path.ts";

/** Whether `p` is a directory (following symlinks), false on any stat error. */
function isDir(p: string): boolean {
  try {
    return statSync(p).isDirectory();
  } catch {
    return false;
  }
}

/** Planctl project directories under the given (or configured) roots. A project
 * is an IMMEDIATE child of a root containing a `.keeper/` data dir. Nested data
 * dirs are NOT surfaced — one level
 * of children per root. Roots are deduplicated, missing / unlistable roots
 * skipped silently. Returned paths are absolute, deduplicated, and sorted.
 * Mirrors discover_projects. */
export function discoverProjects(roots?: string[]): string[] {
  const rootList = roots ?? loadRoots();

  const projects: string[] = [];
  const seen = new Set<string>();
  const seenRoots = new Set<string>();

  for (const root of rootList) {
    if (seenRoots.has(root)) {
      continue;
    }
    seenRoots.add(root);

    let children: string[];
    try {
      children = readdirSync(root)
        .map((name) => join(root, name))
        .sort();
    } catch {
      // Root doesn't exist or can't be listed — skip, not an error.
      continue;
    }

    for (const child of children) {
      if (!isDir(child)) {
        continue;
      }
      if (!hasDataDir(child)) {
        continue;
      }
      if (seen.has(child)) {
        continue;
      }
      seen.add(child);
      projects.push(child);
    }
  }

  return projects.sort();
}

/** Discovered project roots whose data dir holds `tasks/<task_id>.json`. Used
 * by claim to resolve a task's owning project cwd-agnostically. Empty list when
 * no project holds the task (caller maps that to TASK_NOT_FOUND). Mirrors
 * find_projects_with_task. */
export function findProjectsWithTask(
  taskId: string,
  roots?: string[],
): string[] {
  const matches: string[] = [];
  for (const project of discoverProjects(roots)) {
    const dataDir = resolveDataDir(project);
    if (
      dataDir !== null &&
      existsSync(join(dataDir, "tasks", `${taskId}.json`))
    ) {
      matches.push(project);
    }
  }
  return matches;
}

/** Outcome of resolveEpicGlobally — the port of discovery.ResolveResult.
 *
 * Distinguishes the three observable outcomes callers branch on:
 *  - **resolved**: `projectPath` is the owning root, `epicPath` the resolved
 *    data dir's `epics/<id>.json` inside it, `resolvedId` the full slug id
 *    (canonical form for a number-only `fn-N` input). `owners` is empty.
 *  - **not found**: `projectPath`/`epicPath`/`resolvedId` null, `owners` empty.
 *  - **ambiguous**: the id lives in two-or-more projects; all null, `owners`
 *    lists every owner. Callers surface dep_ambiguous_id — never a silent pick.
 */
export class ResolveResult {
  readonly projectPath: string | null;
  readonly epicPath: string | null;
  readonly owners: string[];
  readonly resolvedId: string | null;

  constructor(opts: {
    projectPath?: string | null;
    epicPath?: string | null;
    owners?: string[];
    resolvedId?: string | null;
  }) {
    this.projectPath = opts.projectPath ?? null;
    this.epicPath = opts.epicPath ?? null;
    this.owners = opts.owners ?? [];
    this.resolvedId = opts.resolvedId ?? null;
  }

  /** True iff the id resolved to exactly one project. */
  get resolved(): boolean {
    return this.projectPath !== null;
  }

  /** True iff two-or-more projects carry the id. */
  get ambiguous(): boolean {
    return this.owners.length > 1;
  }
}

/** realpath of `p`, or `p` itself when it can't be resolved (matches the
 * Python `try: p.resolve() except: p` fallback used to dedup the cwd path). */
function realpathOr(p: string): string {
  try {
    return realpathSync(p);
  } catch {
    return p;
  }
}

/** Resolve `epicId` against one project, returning `[fullId, epicPath]` or null.
 * A full-slug input matches its own `<id>.json` exactly (the hot path); a
 * number-only `fn-N` matches the epic whose parsed epic-num equals N by INTEGER
 * equality (never string-prefix: `fn-1` never matches `fn-10`). Mirrors
 * discovery._match_epic_in_project. */
function matchEpicInProject(
  project: string,
  epicId: string,
): [string, string] | null {
  const dataDir = resolveDataDir(project);
  if (dataDir === null) {
    return null;
  }
  const epicsDir = join(dataDir, "epics");

  const exact = join(epicsDir, `${epicId}.json`);
  if (existsSync(exact)) {
    return [epicId, exact];
  }

  const [wantEpic, wantTask] = parseId(epicId);
  if (wantEpic === null || wantTask !== null) {
    return null;
  }
  // Only treat the input as number-only when it has NO slug tail — a slug input
  // that simply doesn't exist must not match a same-number epic.
  if (epicId !== `fn-${wantEpic}`) {
    return null;
  }
  let entries: string[];
  try {
    entries = readdirSync(epicsDir)
      .filter((n) => n.startsWith("fn-") && n.endsWith(".json"))
      .sort();
  } catch {
    return null;
  }
  for (const entry of entries) {
    const stem = entry.slice(0, -".json".length);
    const [candEpic, candTask] = parseId(stem);
    if (candTask !== null) {
      continue;
    }
    if (candEpic === wantEpic) {
      return [stem, join(epicsDir, entry)];
    }
  }
  return null;
}

/** `[projectRoot, fullEpicId]` for every project matching `epicId`. Sibling of
 * findProjectsWithEpic that carries the matched epic's FULL slug id, so a
 * number-only `fn-N` normalizes to canonical form on the write side. Mirrors
 * discovery._find_epic_matches. */
function findEpicMatches(epicId: string, roots?: string[]): [string, string][] {
  const matches: [string, string][] = [];
  for (const project of discoverProjects(roots)) {
    const hit = matchEpicInProject(project, epicId);
    if (hit !== null) {
      matches.push([project, hit[0]]);
    }
  }
  return matches;
}

/** Discovered project roots whose data dir holds `epics/<epicId>.json`.
 * Accepts a number-only `fn-N` (integer-equality) as well as a full slug.
 * Mirrors discovery.find_projects_with_epic. */
export function findProjectsWithEpic(
  epicId: string,
  roots?: string[],
): string[] {
  return findEpicMatches(epicId, roots).map(([project]) => project);
}

/** Resolve an epic id cwd-then-global, distinguishing not-found from ambiguous.
 *
 * Order: (1) cwd short-circuit — if cwd is a planctl project carrying the id, it
 * wins and never counts toward ambiguity; (2) roots discovery — exactly one
 * match resolves, zero is not-found, many is ambiguous (legacy dup state). The
 * roots step is FAIL-SOFT: a discovery that throws or yields nothing contributes
 * no candidates (caller sees not-found, never an exception), while the cwd
 * short-circuit still works without configured roots. Mirrors
 * discovery.resolve_epic_globally. */
export function resolveEpicGlobally(
  epicId: string,
  roots?: string[],
): ResolveResult {
  // 1. Cwd short-circuit — lane-aware. A redirect-eligible worktree lane resolves
  // to its authoritative main STATE repo (the same `classifyCwdVantage` seam the
  // id-less resolver uses), so an id-addressed verb reads/serves the state repo's
  // definition, never the lane's lagging committed snapshot.
  let cwdRoot: string | null;
  try {
    cwdRoot = classifyCwdVantage().effectiveRoot;
  } catch {
    cwdRoot = null;
  }
  if (cwdRoot !== null && hasDataDir(cwdRoot)) {
    const cwdHit = matchEpicInProject(cwdRoot, epicId);
    if (cwdHit !== null) {
      const [resolvedId, epicPath] = cwdHit;
      return new ResolveResult({
        projectPath: cwdRoot,
        epicPath,
        resolvedId,
      });
    }
  }

  // 2. Roots discovery — fail-soft.
  let matches: [string, string][];
  try {
    matches = findEpicMatches(epicId, roots);
  } catch {
    matches = [];
  }

  // Filter the cwd path out so cwd never double-counts as ambiguous.
  if (cwdRoot !== null) {
    const cwdReal = realpathOr(cwdRoot);
    matches = matches.filter(([m]) => realpathOr(m) !== cwdReal);
  }

  if (matches.length === 0) {
    return new ResolveResult({ projectPath: null, epicPath: null });
  }
  if (matches.length === 1) {
    const [owner, resolvedId] = matches[0] as [string, string];
    return new ResolveResult({
      projectPath: owner,
      epicPath: join(
        resolveDataDirOrDefault(owner),
        "epics",
        `${resolvedId}.json`,
      ),
      resolvedId,
    });
  }
  return new ResolveResult({
    projectPath: null,
    epicPath: null,
    owners: matches.map(([m]) => m),
  });
}

// Filename shape for scanEpicIdsGlobal: capture the bare epic id (no ext) of a
// `<id>.json` / `<id>.md` under epics/ or specs/. Mirrors the Python pattern.
const EPIC_FILE_REGEX =
  /^(fn-\d+(?:-[a-z0-9][a-z0-9-]*[a-z0-9]|-[a-z0-9]{1,3})?)\.(json|md)$/;

/** Map every existing bare epic id across `projectPaths` to its owning project.
 * Walks each project's data dir `epics/*` and `specs/fn-*`; a project with no
 * data dir contributes nothing (fail-soft). When the same id appears in multiple
 * projects the LAST-WALKED owner wins — the value feeds human-readable
 * dup-detection messages only, NEVER resolver semantics (use resolveEpicGlobally
 * for that). Mirrors ids.scan_epic_ids_global; listings are sorted at the call
 * site so the last-walked winner is deterministic. */
export function scanEpicIdsGlobal(
  projectPaths: string[],
): Record<string, string> {
  const owners: Record<string, string> = {};
  for (const projectPath of projectPaths) {
    const dataDir = resolveDataDir(projectPath);
    if (dataDir === null) {
      continue;
    }
    for (const sub of ["epics", "specs"]) {
      const subDir = join(dataDir, sub);
      let entries: string[];
      try {
        entries = readdirSync(subDir).sort();
      } catch {
        continue;
      }
      for (const entry of entries) {
        if (!entry.startsWith("fn-")) {
          continue;
        }
        const match = EPIC_FILE_REGEX.exec(entry);
        if (match) {
          owners[match[1] as string] = projectPath;
        }
      }
    }
  }
  return owners;
}

/** The owning project path if ``epicId`` already exists in a DIFFERENT
 * discovered project, else null — the global-name uniqueness check at epic
 * allocation time, so two projects can never mint the same full ``fn-N-slug``
 * even though per-project numbering is independent.
 *
 * Excludes ``localProjectPath`` (a same-project collision is the caller's own
 * epic-path-exists backstop). FAIL-SOFT: if discovery raises or yields no
 * foreign projects, returns null so a fresh / foreign system never hard-breaks
 * creation. Mirrors run_epic_create._check_global_name_unique. */
export function checkGlobalNameUnique(
  epicId: string,
  localProjectPath: string,
): string | null {
  let projects: string[];
  try {
    projects = discoverProjects();
  } catch {
    return null;
  }

  let local: string;
  try {
    local = realpathSync(localProjectPath);
  } catch {
    local = localProjectPath;
  }
  const foreign = projects.filter((p) => {
    let resolved: string;
    try {
      resolved = realpathSync(p);
    } catch {
      resolved = p;
    }
    return resolved !== local;
  });
  if (foreign.length === 0) {
    return null;
  }

  const owners = scanEpicIdsGlobal(foreign);
  return owners[epicId] ?? null;
}
